src/ml/neural_net/combine_base.hpp

```diff
@@ -0,0 +1,253 @@
+   *
+   * A Subscriber is explictly allowed to demand values synchronously from
+   * within its implementation of this method.
+   */
+  virtual void Receive(std::shared_ptr<Subscription> subscription) = 0;
+
+  /**
+   * Transmits a value from the Publisher to this Subscriber.
+   *
+   * A Subcriber should never receive more calls to this method than the total
+   * Demand it has requested from its publisher. Subscribers should only demand
+   * more elements from within this method via its return value.
+   */
+  virtual Demand Receive(Input element) = 0;
+
+  /**
+   * Signals completion of the stream of values from the Publisher.
+   *
+   * A Subscriber should not receive any further signals of any kind after
+   * receiving a Completion.
+   */
+  virtual void Receive(Completion completion) = 0;
+};
+
+/**
+ * Interface for objects that produce values on demand from its Subscribers.
+ *
+ * Unless otherwise specified by the concrete implementation, external
+ * synchronization must be used to avoid concurrent calls on multiple threads to
+ * a Publisher, including via the Subscriptions that it passes to its
+ * Subscribers.
+ *
+ * Each concrete implementation defines whether it is unicast or multicast:
+ * whether multiple Subscribers observe the same values or not. (An
+ * implementation might only support one Subscriber, by passing an immediate
+ * Completion to each Subscriber after the first.)
+ */
+template <typename T>
+class Publisher : public std::enable_shared_from_this<Publisher<T>> {
+ public:
+  /** The type of values that this Publisher produces. */
+  using Output = T;
+
+  virtual ~Publisher() = default;
+
+  /**
+   * Establishes a connection between this Publisher and the given Subcriber.
+   *
+   * The Publisher must eventually call Subscriber::Receive(Subscription) on the
+   * given Subscriber (and may do so synchronously). The Publisher must then
+   * conform to the protocol established by the Subscription.
+   */
+  virtual void Receive(std::shared_ptr<Subscriber<Output>> subscriber) = 0;
+
+  // Convenienience methods, supporting the chaining together of operations.
+  // Many of these rely on the forward declarations above. Client code should
+  // include combine.hpp to ensure these are defined before they are used.
+
+  void Subscribe(std::shared_ptr<Subscriber<Output>> subscriber) {
+    Receive(std::move(subscriber));
+  }
+
+  std::shared_ptr<FuturesStream<Output>> AsFutures() {
+    auto subscriber = std::make_shared<FuturesSubscriber<Output>>();
+    Subscribe(subscriber);
+    return std::make_shared<FuturesStream<Output>>(std::move(subscriber));
+  }
+
+  template <typename TransformType>
+  std::shared_ptr<Publisher<typename TransformType::Output>> Map(
+      std::shared_ptr<TransformType> transform) {
+    using TransformInput = typename TransformType::Input;
+    using TransformOutput = typename TransformType::Output;
+    return std::make_shared<MapPublisher<TransformInput, TransformOutput>>(
```

Comment:
I kept thinking about this weak reference usage. I think there's some trouble with this design or you should state it clearly in the documentation. The Map is only useful when you use it immediately along with the original publisher or the owner is kept living longer than that publisher generated from Map.

```
shared_ptr<Publisher> foo() {
    auto A = make_shared<Publisher>(args);
    ....
    return A->Map(some_transformer_fu);
}

Publisher B = foo();
B->asFutures(); // should throw bad weak_ref because the real owner A expires.
```

Reply:
Talked with @hoytak and I realize I have some misunderstanding of the underlying implementation of `shared_from_this`. The use here is totally legitimate. My mistake.

Reply:
To be fair, I find `shared_from_this` somewhat confusing too. It basically adds an intrusive ref count, instead of relying on the external refcount that `shared_ptr` normally uses. There must be some special logic in all of the `shared_ptr` constructors to detect when it wraps a class that has an intrusive ref count, and then it uses that instead.

Reply:
I finally reasoning it through. Again, if we use this `shared_from_this`,

https://en.cppreference.com/w/cpp/memory/enable_shared_from_this

> it is permitted to call shared_from_this only on a previously shared object, i.e. on an object managed by std::shared_ptr<T>. Otherwise the behavior is undefined (until C++17)std::bad_weak_ptr is thrown (by the shared_ptr constructor from a default-constructed weak_this) (since C++17).

We should let the user know they should first have a shared_ptr to call this class!

> The constructors of std::shared_ptr detect the presence of an unambiguous and accessible (ie. public inheritance is mandatory) (since C++17) enable_shared_from_this base and assign the newly created std::shared_ptr to the internally stored weak reference if not already owned by a live std::shared_ptr (since C++17). 

![image](https://user-images.githubusercontent.com/14238915/74892695-35454a80-533f-11ea-8287-956cfd06a509.png)

I finally understand `owned by a live std::shared_ptr `. @hoytak

Reply:
Makes sense, I'll add a comment!